File: 2255-minimum-swaps-to-group-all-1s-together-ii/2255-minimum-swaps-to-group-all-1s-together-ii.cpp
class Solution {
public:
    int minSwaps(vector<int>& nums) {
        int n = nums.size(), ones = 0, zeros = 0;
        for(int i = 0; i<n; i++) {
            if(nums[i] == 1) ones++;
        }

        for(int i = 0; i<ones; i++) {
            if(nums[i] == 0) zeros++;
        }

        int minSwaps = zeros;
        for (int i = 1; i < n; i++) {
            int start = i;
            int end = (i + ones - 1) % n;

            if (nums[start - 1] == 0) {
                zeros--;
            }
            if (nums[end] == 0) {
                zeros++;
            }

            minSwaps = min(minSwaps, zeros);
        }

        return minSwaps;
    }
};

File: 2255-minimum-swaps-to-group-all-1s-together-ii/2255-minimum-swaps-to-group-all-1s-together-ii.py
class Solution(object):
    def minSwaps(self, nums):
        n = len(nums)
        ones = 0
        zeros = 0

        for i in range(n):
            if nums[i] == 1:
                ones += 1

        for i in range(ones):
            if nums[i] == 0:
                zeros += 1

        min_swaps = zeros
        for i in range(1, n):
            start = i
            end = (i + ones - 1) % n
            if nums[start - 1] == 0:
                zeros -= 1
            if nums[end] == 0:
                zeros += 1
            min_swaps = min(min_swaps, zeros)

        return min_swaps
        